src/mc-fle2-rfds.c

```diff
@@ -0,0 +1,615 @@
+                            mongocrypt_status_t *status)
+{
+   BSON_ASSERT_PARAM (orig);
+   BSON_ASSERT_PARAM (out);
+   BSON_ASSERT (status || true);
+
+   bson_iter_t array, value;
+   const char *op_type_str = bson_iter_key (&in);
+   bool ok = false;
+   const char *field;
+
+   if (!BSON_ITER_HOLDS_ARRAY (&in)) {
+      ERR_WITH_BSON (orig, "%s", "expected argument to be array");
+      goto fail;
+   }
+
+   if (!bson_iter_recurse (&in, &array)) {
+      ERR_WITH_BSON (orig, "%s", "failed to recurse into array");
+      goto fail;
+   }
+
+   // Expect exactly 2 elements, like ["$age", 5]. The first element is the
+   // field path. The second element is the value.
+   if (!bson_iter_next (&array)) {
+      ERR_WITH_BSON (
+         orig, "expected 2 elements in operand %s, got 0", op_type_str);
+      goto fail;
+   }
+   if (!BSON_ITER_HOLDS_UTF8 (&array)) {
+      ERR_WITH_BSON (
+         orig, "expected UTF-8 as first element in %s", op_type_str);
+      goto fail;
+   }
+   field = bson_iter_utf8 (&array, NULL);
+
+   if (!bson_iter_next (&array)) {
+      ERR_WITH_BSON (
+         orig, "expected 2 elements in operand %s, got 1", op_type_str);
+      goto fail;
+   }
+   value = array;
+   if (bson_iter_next (&array)) {
+      ERR_WITH_BSON (
+         orig, "expected 2 elements in operand %s, got > 2", op_type_str);
+      goto fail;
+   }
+
+   out->field = field;
+   out->op_type_str = op_type_str;
+   out->op_type = get_operator_type (op_type_str);
+   out->value = value;
+   ok = true;
+fail:
+   return ok;
+}
+
+// Parses a document like {age: {$gt: 5}}.
+static bool
+parse_match_expression (const bson_t *orig,
+                        bson_iter_t in,
+                        operator_value_t *out,
+                        mongocrypt_status_t *status)
+{
+   BSON_ASSERT_PARAM (orig);
+   BSON_ASSERT_PARAM (out);
+   BSON_ASSERT (status || true);
+
+   bson_iter_t document, value;
+   const char *op_type_str;
+   bool ok = false;
+   const char *field = bson_iter_key (&in);
+
+   if (!BSON_ITER_HOLDS_DOCUMENT (&in)) {
+      ERR_WITH_BSON (orig, "%s", "expected argument to be document");
+      goto fail;
+   }
+
+   if (!bson_iter_recurse (&in, &document)) {
+      ERR_WITH_BSON (orig, "%s", "failed to recurse into document");
+      goto fail;
+   }
+
+   // Expect exactly 1 elements, like {$gt: 5}.
+   if (!bson_iter_next (&document)) {
+      ERR_WITH_BSON (orig, "expected 1 elements in operand %s, got 0", field);
+      goto fail;
+   }
+   op_type_str = bson_iter_key (&document);
+   if (!is_supported_operator (op_type_str)) {
+      ERR_WITH_BSON (orig, "unsupported operator: %s", op_type_str);
+      goto fail;
+   }
+   value = document;
+
+   if (bson_iter_next (&document)) {
+      ERR_WITH_BSON (orig, "expected 1 elements in operand %s, got > 1", field);
+      goto fail;
+   }
+
+   out->field = field;
+   out->op_type_str = op_type_str;
+   out->op_type = get_operator_type (op_type_str);
+   out->value = value;
```

Comment:
```suggestion
   *out = (operator_value_t){
      .field = field,
      .op_type_str = op_type_str,
      .op_type = get_operator_type (op_type_str),
      .value = value,
   };
```
Opt into `-Wmissing-field-initializers` warnings.